In [8]:
library(Rmpfr)
library(float)

genomes=read.csv("genomes.csv")
z = lm(genomes$GC ~ genomes$size)
#head(genomes)
y1=log(genomes$GC)
x1=genomes$size
s = function(x, y){
  mean_x = mean(x)
  mean_y = mean(y)
  nom = sum((x - mean_x)*(y-mean_y))
  denom = sum((x - mean_x)^2)
  m = nom / denom
  return(m)
}
s2 = function(x, y){
  return(cov(x, y)/var(x))
}
intercept = function(x, y, m){
  b = mean(y) - (m * mean(x))
  return(b)
}
R2 = function(x,y){
  x = cor(x,y)^2
  return(x)
  
}
convert = function(z){
    c = z*(10**14)
    return(c)
}

x = convert(x1)
y = convert(y1)


my_size = s(x,y)
my_intercept = intercept(x,y, my_size)
my_R2 = R2(x,y)
print('my_size:')
my_size
print('round:')
round(as.numeric(my_size))
print('przyrównanie do wbudowanej funkcji lm:')
my_size == z$coef[2]
print('z użyciem biblioteki Rmpfr:')
mpfr(my_size, 128)
mpfr(my_size, 333)
print('float')
fl(my_size)
print('z użyciem funkcji sprintf:')       
print(sprintf("%.56f", my_size))

print('my_intercept:')
my_intercept
print('round:')
round(as.numeric(my_intercept))
print('comprassion with lm function:')
my_intercept == z$coef[1]
print('z użyciem biblioteki Rmpfr:')
mpfr(my_intercept, 128)
mpfr(my_intercept, 333)
print('float')
fl(my_intercept)
print('z użyciem funkcji sprintf:')
print(sprintf("%.3f", my_intercept))

#print('my_r2:')
#my_R2

[1] "my_size:"


[1] 0.09099231

[1] "round:"


[1] 0

[1] "przyrównanie do wbudowanej funkcji lm:"


genomes$size 
       FALSE

[1] "z użyciem biblioteki Rmpfr:"


1 'mpfr' number of precision  128   bits 
[1] 0.09099230808282256222074835250168689526618

1 'mpfr' number of precision  333   bits 
[1] 0.09099230808282256222074835250168689526617527008056640625

[1] "float"


# A float32 vector: 1
[1] 0.090992

[1] "z użyciem funkcji sprintf:"
[1] "0.09099230808282256222074835250168689526617527008056640625"
[1] "my_intercept:"


[1] 3.50432e+14

[1] "round:"


[1] 3.50432e+14

[1] "comprassion with lm function:"


(Intercept) 
      FALSE

[1] "z użyciem biblioteki Rmpfr:"


1 'mpfr' number of precision  128   bits 
[1] 350431975154360.875

1 'mpfr' number of precision  333   bits 
[1] 350431975154360.875

[1] "float"


# A float32 vector: 1
[1] 3.5043e+14

[1] "z użyciem funkcji sprintf:"
[1] "350431975154360.875"


Wartość estymatora największa jest dla funkcji mpfr z ustawioną precyzją na 333  oraz funkcji sprintf dla b0:0.09099230808282253446517273687277338467538356781005859375 a dla b1 wartość estymatora jest największa dla funkcji mpfr oraz sprintf: 350431975154360.875

Wartość estymatora najmniejsza jest dla funkcji float dla b0 wynosi: 0.090992 oraz dla b1:3.5043e+14

Różnice pomiędzy wartościami: polegają na zaokrąglaniu liczb,użyte funkcje są bardziej precyzyjne.

Liniowa regresja z użyciem funkcji lm:

In [9]:
df = read.csv('genomes.csv')
head(df)
z = lm(df$GC ~ df$size)
summary(z)

organism                                  group               size GC  
1 Acaryochloris marina MBIC11017            Cyanobacteria       8.36 47.0
2 Acholeplasma laidlawii PG-8A              Firmicutes          1.50 31.9
3 Acidiphilium cryptum JF-5                 Alphaproteobacteria 3.97 67.1
4 Acidithiobacillus ferrooxidans ATCC 53993 Other Bacteria      2.90 58.9
5 Acidothermus cellulolyticus 11B           Actinobacteria      2.40 66.9
6 Acidovorax avenae subsp. citrulli AAC00-1 Betaproteobacteria  5.40 68.5
  habitat     temp.group   temperature
1 Aquatic     Mesophilic   NA         
2 Specialized Mesophilic   37         
3 Multiple    Mesophilic   35         
4 Specialized Mesophilic   NA         
5 Aquatic     Thermophilic 58         
6 Multiple    Mesophilic   NA


Call:
lm(formula = df$GC ~ df$size)

Residuals:
    Min      1Q  Median      3Q     Max 
-28.300  -8.217  -1.238   8.490  28.014 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept)  32.6825     0.8235   39.69   <2e-16 ***
df$size       4.2529     0.2041   20.84   <2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 10.38 on 722 degrees of freedom
Multiple R-squared:  0.3755,	Adjusted R-squared:  0.3746 
F-statistic: 434.1 on 1 and 722 DF,  p-value: < 2.2e-16
